File: and-bible/app/src/main/java/net/bible/android/control/page/ChapterVerse.java
package net.bible.android.control.page;

import net.bible.android.SharedConstants;

import org.crosswire.jsword.passage.Verse;

/**
 * Represent a chapter and verse
 *
 * @author Martin Denham [mjdenham at gmail dot com]
 * @see gnu.lgpl.License for license details.<br>
 * The copyright to this program is held by it's author.
 */
public class ChapterVerse {
	private final int chapter;
	private final int verse;

	public static final ChapterVerse NO_VALUE = new ChapterVerse(SharedConstants.NO_VALUE, SharedConstants.NO_VALUE);

	public ChapterVerse(int chapter, int verse) {
		this.chapter = chapter;
		this.verse = verse;
	}

	public static ChapterVerse fromHtmlId(String chapterDotVerse) {
		String[] strings = chapterDotVerse.split("\\.");
		int chapter = Integer.parseInt(strings[0]);
		int verse = Integer.parseInt(strings[1]);
		return new ChapterVerse(chapter, verse);
	}

	public static ChapterVerse fromVerse(Verse pVerse) {
		int chapter = pVerse.getChapter();
		int verse = pVerse.getVerse();
		return new ChapterVerse(chapter, verse);
	}

	/**
	 * The format used for ids in html
	 */
	public String toHtmlId() {
		return chapter + "." +verse;
	}

	public boolean after(ChapterVerse other) {
		return chapter>other.chapter ||
				(chapter==other.chapter && verse>other.verse);
	}

	public boolean before(ChapterVerse other) {
		return chapter<other.chapter ||
				(chapter==other.chapter && verse<other.verse);
	}

	public boolean sameChapter(ChapterVerse other) {
		return chapter==other.chapter;
	}

	@Override
	public String toString() {
		return "ChapterVerse{" +
				"chapter=" + chapter +
				", verse=" + verse +
				'}';
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;

		ChapterVerse that = (ChapterVerse) o;
		return chapter==that.chapter && verse==that.verse;
	}

	public int getChapter() {
		return chapter;
	}

	public int getVerse() {
		return verse;
	}
}

File: app/src/main/java/net/bible/android/control/page/ChapterVerse.kt
/*
 * Copyright (c) 2020 Martin Denham, Tuomas Airaksinen and the And Bible contributors.
 *
 * This file is part of And Bible (http://github.com/AndBible/and-bible).
 *
 * And Bible is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 *
 * And Bible is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with And Bible.
 * If not, see http://www.gnu.org/licenses/.
 *
 */

package net.bible.android.control.page

import org.crosswire.jsword.passage.Verse

/**
 * Represent a chapter and verse
 *
 * @author Martin Denham [mjdenham at gmail dot com]
 */
data class ChapterVerse(val chapter: Int, val verse: Int) {

    /**
     * The format used for ids in html
     */
    fun toHtmlId(): String = "$chapter.$verse"
    fun toChapterHtmlId(): String = chapter.toString()

    fun after(other: ChapterVerse?): Boolean =
            other != null && (chapter > other.chapter || (chapter == other.chapter && verse > other.verse))

    fun before(other: ChapterVerse?): Boolean =
            other != null && (chapter < other.chapter || (chapter == other.chapter && verse < other.verse))

    fun sameChapter(other: ChapterVerse?): Boolean =
            other != null && (chapter == other.chapter)

    companion object {
        @JvmStatic fun fromHtmlId(chapterDotVerse: String): ChapterVerse {
            val strings = chapterDotVerse.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            val chapter = Integer.parseInt(strings[0])
            val verse = Integer.parseInt(strings[1])
            return ChapterVerse(chapter, verse)
        }

        @JvmStatic fun fromVerse(pVerse: Verse): ChapterVerse {
            val chapter = pVerse.chapter
            val verse = pVerse.verse
            return ChapterVerse(chapter, verse)
        }
    }
}
